snek/snek.py

```diff
@@ -369,61 +506,66 @@ def __build_content(self):
-                    wanted_template = f"{self.config.templates_path}/{page.metadata['template']}"
+            # Determine which template should be used.
+            # If content has a "template" field, check it is valid and use it.
+            template_filepath = self.templates_default
 
-                    # And check if it exists
-                    if os.path.exists(wanted_template):
-                        template_filepath = wanted_template
+            if 'template' in page.metadata and page.metadata['template']:
 
-                #
-                # Render content using template into HTML file
-                #
+                # Append the template folder to the provided template value
+                wanted_template = os.path.join(self.config.templates_path, page.metadata['template'])
 
-                # Parse markdown
-                page.content = markdown.markdown(page.content)
+                # And check if it exists
+                if os.path.exists(wanted_template):
+                    template_filepath = wanted_template
 
-                # Render template
-                lookup = TemplateLookup(directories=[self.config.templates_path,]) # Used to make the template aware of its suroundings
-                renderer = Template(filename=template_filepath, lookup=lookup)
-                html = renderer.render(data=self.data,
-                                       sitemap=self.sitemap,
-                                       config=self.config.__dict__,
-                                       metadata=page.metadata,
-                                       content=page.content)
+            #
+            # Render content using template into HTML file
+            #
 
-                # Write HTML file
-                destination_dirname = os.path.dirname(destination_filepath)
-                if not os.path.exists(destination_dirname):
-                    os.makedirs(destination_dirname)
+            # Parse markdown
+            page.content = markdown.markdown(page.content)
 
-                open(destination_filepath, 'w').write(html)
+            # Render template
+            lookup = TemplateLookup(directories=[self.config.templates_path,]) # Used to make the template aware of its suroundings
+            renderer = Template(filename=template_filepath, lookup=lookup)
+            html = renderer.render(data=self.data,
+                                    sitemap=self.sitemap,
+                                    config=self.config.__dict__,
+                                    metadata=page.metadata,
+                                    content=page.content)
 
-                # Count as built
-                self.pages_built += 1
+            # Write HTML file
+            destination_dirname = os.path.dirname(destination_filepath)
+            if not os.path.exists(destination_dirname):
+                os.makedirs(destination_dirname)
+
+            with open(destination_filepath, 'w') as fp:
```

Comment:
Is there a technical advantage of using `with` in that context over `open(destination_filepath, 'w').write(html)`, given that in the latter the file pointer will also be removed as soon as the operation is done?

Reply:
A bit of OCD on my side here. The file pointer is cleaned up at the end of the execution, so it is potentially left hanging for no good reason. If we have a very large number of files to process / write, that could become a problem. It should actually show up as a warning when running the test suite - I think that's what compelled me to change it :)

Reply:
It is good practice indeed. I _assume_ garbage collection catches it quickly after as we are in a loop context, but better safe than debugging memory leaks ;) .